### ISSUE TITLE
[wordpress] Multiple initializations when replicaCount>1 and using external DB

### ISSUE THREAD

**Issue by vsanjay85** [bitnami/charts#4691]
**Which chart**:
wordpress

**Describe the bug**
When using an external-DB with no pre-existing wordpress data, and 3 compute replicas, the DB is initalized thrice. 

If using the chart-based MariaDB, the PVC access-mode is set to ReadWriteOnce, which probably prevents this from happening. 

However, there is no parameter or config to prevent this from happening with an external DB. 

**To Reproduce**

1. Set wordpress deployment replicaCount to 3
2. Deploy the wordpress helm-chart using a fresh external-DB
3. Open the loadbalancer link a few minutes later.
4. See "Hello World" blog appear thrice, among other duplications.

**Expected behavior**
Regardless of the number of replicas, the external DB is only initialized once.

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
version.BuildInfo{Version:"v3.3.0", GitCommit:"8a4aeec08d67a7b84472007529e8097ec3742105", GitTreeState:"dirty", GoVersion:"go1.14.6"}
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.8", GitCommit:"9f2892aab98fe339f3bd70e3c470144299398ace", GitTreeState:"clean", BuildDate:"2020-08-13T16:12:48Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.9-eks-d1db3c", GitCommit:"d1db3c46e55f95d6a7d3e5578689371318f95ff9", GitTreeState:"clean", BuildDate:"2020-10-20T22:18:07Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```

**Additional context**
Compute pods using an EFS-based PVC

**Comment by dani8art** [comment]
Hi @vsanjay85 

Are you disabling persistence for wordpress pods? I mean adding this parameter:

```
--set persistence.enabled=false
```

**Comment by svaranasi-traderev** [comment]
Hi @dani8art , I am not passing that parameter. As in, I leave it at the default.

**Comment by vsanjay85** [comment]
Hi @dani8art , I am not passing that parameter. As in, I leave it at the default.

**Comment by dani8art** [comment]
If you are not providing it, then it will take the default one and will fail to start, in fact, it claims a `ReadWriteOnce` volume so the pods never get created. 

```yaml
## Enable persistence using Persistent Volume Claims
## ref: http://kubernetes.io/docs/user-guide/persistent-volumes/
##
persistence:
  enabled: true
  ## wordpress data Persistent Volume Storage Class
  ## If defined, storageClassName: <storageClass>
  ## If set to "-", storageClassName: "", which disables dynamic provisioning
  ## If undefined (the default) or set to null, no storageClassName spec is
  ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
  ##   GKE, AWS & OpenStack)
  ##
  # storageClass: "-"
  ##
  ## If you want to reuse an existing claim, you can pass the name of the PVC using
  ## the existingClaim variable
  # existingClaim: your-claim
  accessMode: ReadWriteOnce
  size: 10Gi
```

And you should get this

```console
$ k get pods
NAME                            READY   STATUS              RESTARTS   AGE
wp-mariadb-0                    1/1     Running             0          89m
wp-wordpress-5cdb97748f-d8hpg   0/1     ContainerCreating   0          89m
wp-wordpress-5cdb97748f-fxmpz   1/1     Running             0          89m
wp-wordpress-5cdb97748f-j8cm8   0/1     ContainerCreating   0          89m
```

Could you share more about your specific configuration?

Thanks.

**Comment by vsanjay85** [comment]
I did change that setting to 'ReadWriteMany'. 

Here's the relevant bit:
```
persistence:
  enabled: true
  ## wordpress data Persistent Volume Storage Class
  ## If defined, storageClassName: <storageClass>
  ## If set to "-", storageClassName: "", which disables dynamic provisioning
  ## If undefined (the default) or set to null, no storageClassName spec is
  ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
  ##   GKE, AWS & OpenStack)
  ##
  storageClass: "efs-sc"
  efsId:
  ##
  ## If you want to reuse an existing claim, you can pass the name of the PVC using
  ## the existingClaim variable
  # existingClaim: your-claim
  accessMode: ReadWriteMany
  size: 1Gi
```

where `efs-sc` and the volume come from:

```
---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: efs-pv
spec:
  capacity:
    storage: 1Gi
  volumeMode: Filesystem
  accessModes:
    - {{ .Values.persistence.accessMode | quote }}
  persistentVolumeReclaimPolicy: Retain
  storageClassName: efs-sc
  csi:
    driver: efs.csi.aws.com
    volumeHandle: {{ .Values.persistence.efsId }}

---
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: efs-sc
provisioner: efs.csi.aws.com
```

**Comment by dani8art** [comment]
Hi it seems like there is a race condition between pods in order to initialize these kinds of data, after them are created a file is written to the disk `/bitnami/wordpress/.initialized` you could try something like this if you want to avoid this race condition.

**values.yaml**
```yaml
replicaCount: 3
initContainers: |
  - name: wait-for-initialized
    image: bitnami/minideb:buster
    volumeMounts:
      - mountPath: /bitnami/wordpress
        name: wordpress-data
        subPath: wordpress
    command:
      - /bin/bash
      - -c
      - |
        #!/bin/bash
        if [[ ! -f /bitnami/wordpress/.initialized ]]; then
          if [[ ! -f /bitnami/wordpress/.first_replica ]]; then
            echo $(hostname) > /bitnami/wordpress/.first_replica
          else
            while ! -f /bitnami/wordpress/.initialized; do
              sleep 5s
            done
          fi
        fi
```

**Comment by vsanjay85** [comment]
Thanks @dani8art , that helped.

I already had an initContainer to fix the directory permissions for user:group '1001'. This is what it now looks like:

```
initContainers:
- name: chown-wordpress-mount
  image: bitnami/minideb:buster
  command:
  - /bin/bash
  - -exc
  - |
    set -ex
    chown -R 1001:1001 /bitnami/wordpress
    if [[ ! -f /bitnami/wordpress/.initialized ]]; then
      if [[ ! -f /bitnami/wordpress/.first_replica ]]; then
        echo $(hostname) > /bitnami/wordpress/.first_replica
      else
        while [ ! -f /bitnami/wordpress/.initialized ]; do
          sleep 5s
        done
      fi
    fi
  volumeMounts:
  - name: wordpress-data
    mountPath: /bitnami/wordpress
    subPath: wordpress
```
It would be great if the wordpress initializing script did the check for initializations by itself. However, I'm not sure if/how feasible that is.

**Comment by rafariossaa** [comment]
Hi @vsanjay85 ,
  Thanks for sharing your fix.
  I am opening an internal task to look for a fix for this.

**Comment by marcosbc** [comment]
Hi all, if you are using an external DB, you could try and populate it in a deployment, and then re-deploy afterwards with `skipInstall`. Also FYI, we just released WordPress chart 11.0.0 with [major changes to the container image](https://github.com/bitnami/charts/tree/master/bitnami/wordpress#1100), so it could be worth checking it out.

**Comment by carrodher** [comment]
Unfortunately, this issue was created a long time ago and although there is an internal task to fix it, it was not prioritized as something to address in the short/mid term. It's not a technical reason but something related to the capacity since we're a small team.

Being said that, contributions via PRs are more than welcome in both repositories (containers and charts). Just in case you would like to contribute.

During this time, there are several releases of this asset and it's possible the issue has gone as part of other changes. If that's not the case and you are still experiencing this issue, please feel free to reopen it and we will re-evaluate it.
